compiler/noirc_frontend/src/parser/parser.rs

```diff
@@ -1101,21 +1102,69 @@ where
         .labelled(ParsingRuleLabel::Atom)
 }
 
-fn quote<'a, P>(statement: P) -> impl NoirParser<ExpressionKind> + 'a
-where
-    P: NoirParser<StatementKind> + 'a,
-{
-    keyword(Keyword::Quote).ignore_then(spanned(block(statement))).validate(
-        |(block, block_span), span, emit| {
+fn quote() -> impl NoirParser<ExpressionKind> {
+    keyword(Keyword::Quote).ignore_then(spanned(token_stream_block(false, true))).validate(
+        |(tokens, block_span), span, emit| {
             emit(ParserError::with_reason(
                 ParserErrorReason::ExperimentalFeature("quoted expressions"),
                 span,
             ));
-            ExpressionKind::Quote(block, block_span)
+            ExpressionKind::Quote(Tokens(tokens), block_span)
         },
     )
 }
 
+/// Parses a stream of tokens terminated by '{' or '}'.
+/// - parse_braces: if true, parses '{' and '}' surrounding the token stream.
+/// - include_braces: if true, include the surrounding braces in the returned tokens vec
+fn token_stream_block(
+    include_braces: bool,
+    parse_braces: bool,
+) -> impl NoirParser<Vec<SpannedToken>> {
+    let append_vecs = |(mut vec1, mut vec2): (Vec<_>, _)| {
+        vec1.append(&mut vec2);
+        vec1
+    };
+
+    // Parse a stream of tokens ending in '{' or '}'.
+    // - If we ended with a '}': end
+    // - If we ended with a '{': recursively parse another token stream block
+    let inner_stream = spanned(none_of([Token::LeftBrace, Token::RightBrace]))
+        .map(|(token, span)| SpannedToken::new(token, span))
+        .repeated()
+        .then(one_of([Token::LeftBrace, Token::RightBrace]).rewind().then_with(move |end| {
+            match end {
+                Token::LeftBrace => token_stream_block(true, true)
+                    .then(token_stream_block(false, false))
```

Comment:
Can you add a test with 3+ nesting levels?

Reply:
Good idea - added. Writing the parser really shouldn't have been the most difficult part of this PR...